include/GafferScene/PrimitiveSampler.h

```diff
@@ -0,0 +1,131 @@
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERSCENE_PRIMITIVESAMPLER_H
+#define GAFFERSCENE_PRIMITIVESAMPLER_H
+
+#include "GafferScene/Deformer.h"
+
+#include "Gaffer/StringPlug.h"
+
+#include "IECoreScene/PrimitiveEvaluator.h"
+
+namespace GafferScene
+{
+
+/// Base class for nodes which use an `IECoreScene::PrimitiveEvaluator`
+/// to sample primitive variables from another object.
+class GAFFERSCENE_API PrimitiveSampler : public Deformer
+{
+
+	public :
+
+		~PrimitiveSampler() override;
+
+		GAFFER_GRAPHCOMPONENT_DECLARE_TYPE( GafferScene::PrimitiveSampler, PrimitiveSamplerTypeId, Deformer );
+
+		ScenePlug *sourcePlug();
+		const ScenePlug *sourcePlug() const;
+
+		Gaffer::StringPlug *sourceLocationPlug();
+		const Gaffer::StringPlug *sourceLocationPlug() const;
+
+		Gaffer::StringPlug *primitiveVariablesPlug();
+		const Gaffer::StringPlug *primitiveVariablesPlug() const;
+
+		Gaffer::StringPlug *prefixPlug();
+		const Gaffer::StringPlug *prefixPlug() const;
+
+		Gaffer::StringPlug *statusPlug();
+		const Gaffer::StringPlug *statusPlug() const;
+
+	protected :
+
+		PrimitiveSampler( const std::string &name = defaultName<PrimitiveSampler>() );
+
+		/// SamplingFunction
+		/// ================
+		///
+		/// Derived classes are responsible for generating a `SamplingFunction`, which
+		/// performs an `IECoreScene::PrimitiveEvaluator` query for a single index within the
+		/// input primitive. The base class takes care of everything else.
+
+		using SamplingFunction = std::function<bool (
+			/// The PrimitiveEvaluator to use for sampling.
+			const IECoreScene::PrimitiveEvaluator &evaluator,
+			/// The index within the input to date to sample for.
+			size_t index,
+			/// A transform that must be applied to any geometric
+			/// data before querying the `evaluator`. This converts
+			/// from the object space of the sampling object into
+			/// the object space of the source object.
+			const Imath::M44f &transform,
+			/// The destination for the result of the `evaluator`
+			/// query.
+			IECoreScene::PrimitiveEvaluator::Result &result
+		)>;
+
+		/// Must be implemented to return true if the specified plug affects the
+		/// generation of the `SamplingFunction`. All implementations should call
+		/// the base implementation first, and return `true` if it does.
+		virtual bool affectsSamplingFunction( const Gaffer::Plug *input ) const = 0;
+		/// Must be implemented to hash all plugs that are used in `computeSamplingFunction()`.
+		/// All implementation should call the base class implementation first.
+		virtual void hashSamplingFunction( IECore::MurmurHash &h ) const = 0;
+		/// Must be implemented to return an appropriate `SamplingFunction` for the
+		/// specified primitive, and fill `interpolation` with the interpolation of
+		/// the primitive variables being used for the query. The sampling function
+		/// will then be queried with `index` values in the interval
```

Comment:
Sorry if I have my wires crossed, but is it worth clarifying that `primitive` is the _input_ primitive not the source one? Once you're fully up to speed with the evaluator semantics, then its apparent that this is the case, but from a high level  it might be misconstrued as being about the variables being sampled, rather than the variables used to drive the sampling 'point'.

Reply:
I did struggle with the naming a bit here. But I wanted a name more distinct than _input_, because  the source is also an input. I tried "sampling object" in the `SamplingFunction` docs, but that didn't feel great either so I sheepishly only used it there. I'm open to ideas for better and more consistent naming, but I'm not sure _input_ is it. And after all, you did get your wires completely parallel even with the comments as they are :)

Reply:
Hehe. It took a lot of cross-referencing and sanity checking though :) How about something along the lines of the following. Don't know if its any better:

```
/// Must be implemented to return an appropriate `SamplingFunction` for
/// the destination primitive, and fill `interpolation` with the interpolation of
/// the resultant output primitive variables. The sampling function will then
/// be queried with `index` values in the interval
///  `[ 0, destinationPrimitive->variableSize( interpolation ) )`.
```

Reply:
Took a stab at something like this in 6d4f262, updating some other comments and variable names to match.

Reply:
Spot on 👌. Thanks John, think that really helps explain it to newcomers who may be wanting to implement their own `PrimitiveSampler` derived node.